### PR TITLE
Update to Yarn3

### PR DESCRIPTION
**Pull request by rurikoaraki**
### Platforms Impacted
- [ ] iOS
- [ ] macOS
- [ ] win32 (Office)
- [ ] windows
- [ ] android

### Description of changes

Updating repo to use yarn 3, per instructions [here](https://yarnpkg.com/getting-started/migration).

1. Fix the use-yarn-please script to look at the new places
2. Various yarn-related checkins
3. Add scripts package as a dependency for packages missing it
4. Fix instance of ES2019 usage, which we don't support
5. Add plugin for disabling patching of TS
6. Fix workspaces property in package.json
7. Port tester_deps

### Verification

Run yarn, run yarn build, run yarn test

### Pull request checklist

This PR has considered (when applicable):
- [ ] Automated Tests
- [ ] Documentation and examples
- [ ] Keyboard Accessibility
- [ ] Voiceover
- [ ] Internationalization and Right-to-left Layouts

### CODE REVIEW

**Review by Saadnajmi — APPROVED** [review]

File: scripts/use-yarn-please.js [hunk 1]
```diff
@@ -60,6 +60,12 @@ function checkRepositoryLocation() {
 }
 
 function detectYarnInstallation() {
-  const yarnResult = spawnSync('yarn', ['--version']);
+  let yarnResult;
+  if (process.platform !== 'win32') {
+    yarnResult = spawnSync('yarn.cmd', ['--version']);
+  } else {
```
Comment (Saadnajmi) [inline scripts/use-yarn-please.js:66]: Shouldn't this be if it _is_ win32?

Comment (rurikoaraki) [reply scripts/use-yarn-please.js:66]: Not entirely sure. This file works for me as it is now, maybe I asked you to try it out the first time around? I just copied this from https://github.com/microsoft/fluentui-react-native/pull/2054

File: .yarnrc.yml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+nodeLinker: node-modules
```
Comment (tido64) [inline .yarnrc.yml:1]: You should copy the plugin for disabling PnP patching of TypeScript and others: https://github.com/microsoft/rnx-kit/pull/2315

File: scripts/use-yarn-please.js (hunk 1 above)
Comment (tido64) [reply scripts/use-yarn-please.js:66]: `.cmd` is Windows only. I think the check was inverted.

Comment (rurikoaraki) [reply scripts/use-yarn-please.js]: Got it, fixed

**Comment by rurikoaraki** [comment]
@tido64 Would you happen to know why install-windows-test-app is erroring out?

**Comment by tido64** [comment]
This looks like a bug in RNTA. `node_modules` is no longer created under `/apps/fluent-tester`, so it generates Windows project files under `/node_modules` instead while `app.json` assumes `/apps/fluent-tester/node_modules`. As a workaround, you can do this:

```diff
diff --git a/apps/fluent-tester/react-native.config.js b/apps/fluent-tester/react-native.config.js
index 1bae4e2a6..12ee75eea 100644
--- a/apps/fluent-tester/react-native.config.js
+++ b/apps/fluent-tester/react-native.config.js
@@ -1,7 +1,7 @@
 const project = (() => {
   try {
     const { configureProjects } = require('react-native-test-app');
-    return configureProjects({
+    const project = configureProjects({
       android: {
         sourceDir: 'android',
       },
@@ -13,6 +13,11 @@ const project = (() => {
         solutionFile: 'windows/FluentTester.sln',
       },
     });
+    // FIXME: This is a bug in RNTA
+    if (project.windows) {
+      project.windows.project.projectFile = '..\\..\\' + project.windows.project.projectFile;
+    }
+    return project;
   } catch (_) {
     return undefined;
   }

```

We should fix this as well:

```
➤ YN0050: The --frozen-lockfile option is deprecated; use --immutable and/or --immutable-cache instead
```

With Yarn v3, immutable is enabled by default if CI is detected. We should be able to just run `yarn`.

**Comment by rurikoaraki** [comment]
> This looks like a bug in RNTA. `node_modules` is no longer created under `/apps/fluent-tester`, so it generates Windows project files under `/node_modules` instead while `app.json` assumes `/apps/fluent-tester/node_modules`. As a workaround, you can do this:
> 
> ```diff
> diff --git a/apps/fluent-tester/react-native.config.js b/apps/fluent-tester/react-native.config.js
> index 1bae4e2a6..12ee75eea 100644
> --- a/apps/fluent-tester/react-native.config.js
> +++ b/apps/fluent-tester/react-native.config.js
> @@ -1,7 +1,7 @@
>  const project = (() => {
>    try {
>      const { configureProjects } = require('react-native-test-app');
> -    return configureProjects({
> +    const project = configureProjects({
>        android: {
>          sourceDir: 'android',
>        },
> @@ -13,6 +13,11 @@ const project = (() => {
>          solutionFile: 'windows/FluentTester.sln',
>        },
>      });
> +    // FIXME: https://github.com/microsoft/react-native-test-app/issues/1529
> +    if (project.windows) {
> +      project.windows.project.projectFile = '..\\..\\' + project.windows.project.projectFile;
> +    }
> +    return project;
>    } catch (_) {
>      return undefined;
>    }
> ```
> 
> We should fix this as well:
> 
> ```
> ➤ YN0050: The --frozen-lockfile option is deprecated; use --immutable and/or --immutable-cache instead
> ```
> 
> With Yarn v3, immutable is enabled by default if CI is detected. We should be able to just run `yarn`.
> 
> Update: [microsoft/react-native-test-app#1530](https://github.com/microsoft/react-native-test-app/pull/1530)

Thanks @tido64! I pulled in your fix

File: .yarnrc.yml (hunk 2 above)
Comment (rurikoaraki) [reply .yarnrc.yml:1]: Done

**Comment by rurikoaraki** [comment]
@tido64 I'm going to merge this since it gets conflicts every time yarn.lock changes, but feel free to leave feedback if you have any other comments

**Pull request merged**